File: Sketches/MH/Inversion/miniaxon.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2010 British Broadcasting Corporation and Kamaelia Contributors(1)
#
# (1) Kamaelia Contributors are listed in the AUTHORS file and at
#     http://www.kamaelia.org/AUTHORS - please extend this file,
#     not this notice.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import Queue
import threading
import time

class microprocess(object):
    def __init__(self):
        super(microprocess, self).__init__()
    def main(self):
        yield 1
    def activate(self, scheduler,mainmethod="main"):
        self._mprocess = self.__getattribute__(mainmethod)()
        self.scheduler = scheduler
        self.scheduler.addThread(self)
        return self
    def next(self):
        return self._mprocess.next()


class scheduler(microprocess):
    def __init__(self):
        super(scheduler, self).__init__()
        self.active = []
        self.newqueue = []
        self.event = threading.Event()
    def notify(self):
        self.event.set()
    def main(self): 
        while len(self.newqueue):
            self.active = self.newqueue
            self.newqueue = []
            activityCount = 0
            waitingCount = 0
            for current in self.active:
                yield 1
                try:
                    result = current.next()
                    if result is not -1:
                        activityCount +=1 
                        self.newqueue.append(current)
                        if result=="BLOCK" or current.paused:
                            waitingCount += 1
                except StopIteration:
                    pass
            if activityCount > 0 and waitingCount == activityCount:
                self.event.wait()
                self.event.clear()
    def addThread(self, thread):
        self.newqueue.append(thread)
        
class box(list):
    def __init__(self, notify):
        self.notify=notify
        super(box,self).__init__()
    def append(self, value):
        super(box,self).append(value)
        self.notify()

class component(microprocess):
    def __init__(self):
        super(component, self).__init__()
        self.boxes = { "inbox" : box(self.notify), "outbox": box(self.notify) }
        self.paused=False
    def send(self, value, outboxname):
        self.boxes[outboxname].append(value)
    def recv(self, inboxname):
        result = self.boxes[inboxname][0]
        del self.boxes[inboxname][0]
        return result
    def dataReady(self, inboxname):
        return len(self.boxes[inboxname])
    def link(self, src, dst):
        scomp,sbox = src
        dcomp,dbox = dst
        while len(scomp.boxes[sbox]):
            dcomp.boxes[dbox].append(scomp.boxes[sbox].pop(0))
        scomp.boxes[sbox] = dcomp.boxes[dbox]
        return (src, dst)
    def unlink( linkage ):
        ((scomp,sbox),(dcomp,dbox)) = linkage
        scomp.boxes[sbox] = box(self.notify)
    def notify(self):
        self.paused=False
    def pause(self):
        self.paused=True

class threadedcomponent(component):
    """Very basic threaded component, with poor thread safety"""
    def __init__(self):
        super(threadedcomponent,self).__init__()
        self.queues = {}
        self.isInbox = { "inbox":1, "outbox":0 }
        for box in self.boxes:
            self.queues[box] = Queue.Queue()
    def activate(self,scheduler):
        return component.activate(self, scheduler, "_localmain")
    def send(self, value, outboxname):
        self.queues[outboxname].put(value)
        self.scheduler.notify()
    def recv(self, inboxname):
        return self.queues[inboxname].get()
    def dataReady(self, inboxname):
        return self.queues[inboxname].qsize()
    
    def _localmain(self):
        self._thread = threading.Thread(target=self._threadrun)
        self._thread.setDaemon(True)
        self._threadalive=True
        self._thread.start()
        while self._threadalive and self._thread.isAlive():
            yield 1
            for boxname in self.queues:
                if self.isInbox[boxname]:
                    while component.dataReady(self,boxname):
                        self.queues[boxname].put( component.recv(self,boxname) )
                else:
                    while not self.queues[boxname].empty():
                        component.send(self, self.queues[boxname].get(), boxname)
            self.pause()

    def _threadrun(self):
        self.main()
        self._threadAlive=False
        self.scheduler.notify()
        
##  events!!!
# --------------------------------------------------

class Producer(threadedcomponent):
    def __init__(self, sleeptime=0.2):
        super(Producer,self).__init__()
        self.sleeptime=sleeptime
    def main(self):
        for i in range(10):
            time.sleep(self.sleeptime)
            self.send(i,"outbox")
        self.send("DONE","outbox")

class Output(component):
    def __init__(self,name):
        super(Output,self).__init__()
        self.name=name
    def main(self):
        done=False
        while not done:
            while self.dataReady("inbox"):
                msg=self.recv("inbox")
                print self.name+" "+str(msg)
                done = done or (msg == "DONE")
            if not done:
                self.pause()
                yield 1
                print self.name+" reawoken..."

sched=scheduler()
p=Producer().activate(sched)
o=Output("A").activate(sched)

p.link( (p,"outbox"),(o,"inbox") )

# shouldn't affect first outputter
p2=Producer(0.4).activate(sched)
o2=Output("                B").activate(sched)

p.link( (p2,"outbox"),(o2,"inbox") )

for _ in sched.main():
    pass

File: Code/CPP/Scratch/miniaxon.cpp
/*
 * Copyright 2010 British Broadcasting Corporation and Kamaelia Contributors(1)
 *
 * (1) Kamaelia Contributors are listed in the AUTHORS file and at
 *     http://www.kamaelia.org/AUTHORS - please extend this file,
 *     not this notice.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * -------------------------------------------------------------------------
 *
 * This file contains a microcosm proof of concept implementation of a
 * subset of Axon & a simple producer consumer system using this.
 *
 * Compile:
 *  > g++ -o miniaxon miniaxon.cpp
 *
 * Run:
 *  > ./miniaxon
 *
 *
 * This is a rather icky implementation of Axon now, which has many of the
 * normal features of axon - the ability to link arbitrary outboxes to
 * arbitrary inboxes, multiple named inboxes/outboxes, arbitrary message
 * passing. (Subclass IPC_S, and place what you want in a .payload
 * attribute. (doesn't have to have a .payload though).
 *
 */

#include "generators.hpp"
#include <iostream>
#include <list>
#include <map>

using namespace std;

/* Base class for IPC objects */
struct IPC_S {
   virtual string __str__() { }; // Force polymorphic
   int payload;
};

struct Microprocess : public Generator {
    Microprocess()  {     };
    ~Microprocess() {     };
    virtual int next() {
    GENERATOR_CODE_START
        cout << "X Hello!" << endl;
        YIELD(-1);
    GENERATOR_CODE_END
    };
    virtual void run() {
    // This would not normally be run on a microprocess, but may be useful
    // for some circumstances, such as testing.
       int r;
       while(r!= -1) {
          r = next();
       }
    }
};


struct Scheduler : public Microprocess {
    list<Microprocess*> active;
    list<Microprocess*> newqueue;
    list<Microprocess*>::iterator current;

    int i;
    virtual int next() {
    GENERATOR_CODE_START
        i = 0;
        YIELD(1);
        for(i=0; i< 100; i++) {
           newqueue.resize(0);
           for(current=active.begin(); current != active.end(); current++) {
               int result;
               try {
                  result = (*current)->next();
                  if (result != -1) {
                     newqueue.push_back(*current);
                  };
               } catch (StopIteration null){
                  int k;
                  k =1;
               };
           }
           active.resize(0);  // Empty the active queue (unnecessary?)
           active = newqueue; // _Copies_ vector newqueue to active
           YIELD(1);
        };
        YIELD(-1);
    GENERATOR_CODE_END
    };
    void activateMicroprocess(Microprocess* m) {
        active.push_back(m);
    };
};


/* Simplificaton of Component, need to check what it's now missing
 */
struct SimpleComponent : public Microprocess {
    map<string,list<IPC_S*> > outboxes;
    map <string, list<IPC_S*> > inboxes;

    SimpleComponent()  {
       inboxes["inbox"];
       outboxes["outbox"];
    };
    ~SimpleComponent() {     };

    virtual int next() {
    GENERATOR_CODE_START
        cout << "SimpleComponent" << endl;
        YIELD(-1);
    GENERATOR_CODE_END
    };

    void send(IPC_S *value) {   send(value,"outbox"); }
    void send(IPC_S *value, string somebox) {
       outboxes[somebox].push_back(value);
    }

    IPC_S *recv() {   return recv("inbox"); }
    IPC_S *recv(string somebox) {
        IPC_S *result;
        result = *(inboxes[somebox].begin());
        inboxes[somebox].erase(inboxes[somebox].begin());
        return result;
    }

    void deliver(IPC_S *value, string somebox) {
       inboxes[somebox].push_back(value);
    }
    IPC_S *collect(string somebox) {
        IPC_S *result;
        result = *(outboxes[somebox].begin());
        outboxes[somebox].erase(outboxes[somebox].begin());
        return result;
    }
    bool dataReady() { return dataReady("inbox");    }
    bool dataReady(string somebox) {
       return ! inboxes[somebox].empty();
    }
    bool dataOutReady(string somebox) {
       return ! outboxes[somebox].empty();
    }
};

struct Postman : public SimpleComponent {
    SimpleComponent* source;
    string sourcebox;
    SimpleComponent* destination;
    string destbox;
    IPC_S * temp;
    Postman(SimpleComponent* s, SimpleComponent* d)  {
       source = s;
       destination = d;
       sourcebox = "outbox";
       destbox = "inbox";
    };
    Postman(SimpleComponent* s, string sbox, SimpleComponent* d, string dbox)  {
       source = s;
       destination = d;
       sourcebox = sbox;
       destbox = dbox;
    };
    ~Postman() {     };

    virtual int next() {
    GENERATOR_CODE_START
        while (1) {
           if (source->dataOutReady(sourcebox)) {
              temp = source->collect(sourcebox);
              destination->deliver(temp,destbox);
           };
           YIELD(1);
        };
    GENERATOR_CODE_END
    };
};

/**********************************************************************
 *
 * Sample Producer/Consumer system, passing IPC messages with a string as
 * payload.
 *
 */

struct mystring : public IPC_S{
   string payload;
};

struct Producer : public SimpleComponent {

    mystring msg;
    Producer()  {     };
    ~Producer() {     };

    virtual int next() {
    GENERATOR_CODE_START
        while(1) {
           msg.payload = "hello world!";
           send(&msg);
           YIELD(1);
        };
    GENERATOR_CODE_END
    };
};

struct Consumer : public SimpleComponent {
    mystring *result;

    Consumer()  {     };
    ~Consumer() {     };

    virtual int next() {
    GENERATOR_CODE_START
        while (1) {
            if (dataReady()) {
               result = dynamic_cast<mystring *>(recv());
               cout << "! " << result->payload << endl;
            };
            YIELD(1);
        };
    GENERATOR_CODE_END
    };
};

/*
 * End of sample system.
 *
 **********************************************************************/

int main(int, char **) {
   Scheduler scheduler;

   Producer P;
   Consumer C;
   Postman postie(&P,"outbox", &C, "inbox"); // Same as postie(&P,&C)
   scheduler.activateMicroprocess(&P);
   scheduler.activateMicroprocess(&C);
   scheduler.activateMicroprocess(&postie);
   scheduler.run();

   return 0;
}
